instrumentation/mongo/src/main/java/brave/mongo/TraceMongoCommandListener.java

```diff
@@ -0,0 +1,163 @@
+
+    collectionName.ifPresent(collection -> span.tag("mongo.collection", collection));
+
+    final InetSocketAddress socketAddress = event.getConnectionDescription().getServerAddress().getSocketAddress();
+    span.remoteIpAndPort(socketAddress.getAddress().getHostAddress(), socketAddress.getPort());
+
+    span.start();
+  }
+
+  @Override public void commandSucceeded(final CommandSucceededEvent event) {
+    final Span span = activeSpans.remove(event.getRequestId());
+    if (span == null) {
+      return;
+    }
+    span.finish();
+  }
+
+  @Override public void commandFailed(final CommandFailedEvent event) {
+    final Span span = activeSpans.remove(event.getRequestId());
+    if (span == null) {
+      return;
+    }
+    span.error(event.getThrowable());
+    span.finish();
+  }
+
+  static Optional<String> getCollectionName(final BsonDocument command) {
+    final Iterator<Map.Entry<String, BsonValue>> iterator = command.entrySet().iterator();
+    if (!iterator.hasNext()) {
+      return Optional.empty();
+    }
+    final BsonValue value = iterator.next().getValue();
+    if (!value.isString()) {
+      return Optional.empty();
+    }
+    final String stringValue = value.asString().getValue().trim();
+    return stringValue.isEmpty() ? Optional.empty() : Optional.of(stringValue);
+  }
+
+  /**
+   * Returns an abbreviated version of the command for logging/tracing purposes. Currently this is simply a
+   * truncated version of the command's JSON string representation.
+   *
+   * Note that sensitive data related to the Mongo protocol itself is already scrubbed at this point according to
+   * https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#security
+   *
+   * @return an abbreviated version of the command for logging/tracing purposes
+   */
+  Optional<String> getAbbreviatedCommand(final BsonDocument command) {
+    if (maxAbbreviatedCommandLength <= 0) {
+      return Optional.empty();
+    }
+    final String abbreviatedCommand = command.toJson(jsonWriterSettings);
+    return abbreviatedCommand.isEmpty() ? Optional.empty() : Optional.of(abbreviatedCommand);
+  }
+
+  static String getSpanName(final String commandName, final Optional<String> collectionName) {
+    return collectionName.map(collection -> commandName + " " + collection).orElse(commandName);
+  }
+
+  public static class Builder {
+    private Tracer tracer;
+    private int maxAbbreviatedCommandLength = 100;
+
+    /**
+     * Sets the {@link Tracer} to use. Required.
+     */
+    public Builder tracer(final Tracer tracer) {
+      this.tracer = tracer;
+      return this;
+    }
+
+    /**
+     * Sets how many characters of the Mongo command to report in the "mongo.command" tag. Defaults to 100.
+     * If non-positive, command reporting will be disabled.
+     */
+    public Builder maxAbbreviatedCommandLength(final int maxAbbreviatedCommandLength) {
```

Comment:
Better to validate and not accept negative values (have no real meaning and more likely a user bug) instead of using `Math.max` above.